File: snakes/go/spring-league-2022/logic.go
package main

// This file can be a nice home for your Battlesnake logic and related helper functions.
//
// We have started this for you, with a function to help remove the 'neck' direction
// from the list of possible moves!

import (
	"log"
	"math/rand"
	"sort"
)

// This function is called when you register your Battlesnake on play.battlesnake.com
// See https://docs.battlesnake.com/guides/getting-started#step-4-register-your-battlesnake
// It controls your Battlesnake appearance and author permissions.
// For customization options, see https://docs.battlesnake.com/references/personalization
// TIP: If you open your Battlesnake URL in browser you should see this data.
func info() BattlesnakeInfoResponse {
	log.Println("INFO")
	return BattlesnakeInfoResponse{
		APIVersion: "1",
		Author:     "ES Team", // TODO: Your Battlesnake username
		Color:      "#888887", // TODO: Personalize
		Head:       "gamer",   // TODO: Personalize
		Tail:       "gamer", // TODO: Personalize
	}
}

// This function is called everytime your Battlesnake is entered into a game.
// The provided GameState contains information about the game that's about to be played.
// It's purely for informational purposes, you don't have to make any decisions here.
func start(state GameState) {
	log.Printf("%s START\n", sanatizeInput(state.Game.ID))
}

// This function is called when a game your Battlesnake was in has ended.
// It's purely for informational purposes, you don't have to make any decisions here.
func end(state GameState) {
	log.Printf("%s END\n\n", sanatizeInput(state.Game.ID))
}

// This function is used to check if a coordinate is in hazards.
func isHazard(x int, y int, hazards []Coord) bool {
	for _, hazard := range hazards {
		if hazard.X == x && hazard.Y == y {
			return true
		}
	}
	return false
}

//This function is used to check if a coordinate is in your snake body.
func isBody(x int, y int, body []Coord) bool {
	for _, coord := range body {
		if coord.X == x && coord.Y == y {
			return true
		}
	}
	return false
}

// This function is used to check if another snake is in the coordiantes.
func isSnake(x int, y int, snakes []Battlesnake) bool {
	for _, snake := range snakes {
		for _, coord := range snake.Body {
			if coord.X == x && coord.Y == y {
				return true
			}
		}
	}
	return false
}

// This function is used to check if our head x,y coord is on the edge of the board.
func onEdge(x int, y int, width int, height int) bool {
	if x == 0 || x == width-1 || y == 0 || y == height-1 {
		return true
	}
	return false
}

// Function that takes in possibleMoves and tells us the currently available safe moves.
func safeMoves(possibleMoves map[string]bool) []string {
	var moves []string
	for move, isSafe := range possibleMoves {
		if isSafe {
			moves = append(moves, move)
		}
	}
	return moves
}

// create abs to use with sort.
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// Sort the food coordinates by manhattan distance from the head.
func sortFood(state GameState) []Coord {
	var food []Coord
	for _, f := range state.Board.Food {
		food = append(food, f)
	}
	sort.Slice(food, func(i, j int) bool {
		dist1 := abs(food[i].X-state.You.Body[0].X) + abs(food[i].Y-state.You.Body[0].Y)
		dist2 := abs(food[j].X-state.You.Body[0].X) + abs(food[j].Y-state.You.Body[0].Y)
		return dist1 < dist2
	})
	return food
}

// This function is called on every turn of a game. Use the provided GameState to decide
// where to move -- valid moves are "up", "down", "left", or "right".
// We've provided some code and comments to get you started.
func move(state GameState) BattlesnakeMoveResponse {
	possibleMoves := map[string]bool{
		"up":    true,
		"down":  true,
		"left":  true,
		"right": true,
	}

	// Don't move back on our own neck
	myHead := state.You.Body[0] // Coordinates of your head
	myNeck := state.You.Body[1] // Coordinates of body piece directly behind your head (your "neck")
	if myNeck.X < myHead.X {
		possibleMoves["left"] = false
	} else if myNeck.X > myHead.X {
		possibleMoves["right"] = false
	} else if myNeck.Y < myHead.Y {
		possibleMoves["down"] = false
	} else if myNeck.Y > myHead.Y {
		possibleMoves["up"] = false
	}

	// Avoid walls if we are in a game mode that doesn't allow wrapping.
	boardWidth := state.Board.Width
	boardHeight := state.Board.Height
	gameMode := state.Game.Ruleset.Name
	// Move away from edge of board if we are not in wrapped game mode.
	if gameMode != "wrapped" {
		if myHead.X == 0 {
			log.Println("We are at the left edge of the board")
			possibleMoves["left"] = false
		}
		if myHead.X == boardWidth-1 {
			log.Println("We are at the right edge of the board")
			possibleMoves["right"] = false
		}
		if myHead.Y == 0 {
			log.Println("We are at the top edge of the board")
			possibleMoves["down"] = false
		}
		if myHead.Y == boardHeight-1 {
			log.Println("We are at the bottom edge of the board")
			possibleMoves["up"] = false
		}
	}

	// Don't hit yourself.
	mybody := state.You.Body
	// If a body part is immediately to the right or left of your head, you can't move right or left.
	// Check if a body part is to the right of head. If it is, then we can't move right.
	if isBody(myHead.X+1, myHead.Y, mybody) {
		possibleMoves["right"] = false
	}
	// Check to see if body part is to the left of head. If it is, then we can't move left.
	if isBody(myHead.X-1, myHead.Y, mybody) {
		possibleMoves["left"] = false
	}
	// If a body part is immediately above or below your head, you can't move up or down.
	// Check to see if body part is above head. If it is, then we can't move up.
	if isBody(myHead.X, myHead.Y+1, mybody) {
		possibleMoves["up"] = false
	}
	// Check to see if body part is below head. If it is, then we can't move down.
	if isBody(myHead.X, myHead.Y-1, mybody) {
		possibleMoves["down"] = false
	}
	// If we are in the wrapped game mode and we are on the edge of the board, we need to avoid wrapping into our own body.
	if gameMode == "wrapped" && onEdge(myHead.X, myHead.Y, boardWidth, boardHeight) {
		// Print that we are in this section of the code for debugging purposes.
		log.Println("We are in wrapped game mode and don't want to wrap into our own body")
		// If our head is at x = boardWidth - 1 then check if isBody(0, y) is true.
		if myHead.X == boardWidth-1 && isBody(0, myHead.Y, mybody) {
			possibleMoves["right"] = false
		}
		// If our head is at x = 0 then check if isBody(boardWidth-1, y) is true.
		if myHead.X == 0 && isBody(boardWidth-1, myHead.Y, mybody) {
			possibleMoves["left"] = false
		}
		// If our head is at y = boardHeight - 1 then check if isBody(x, 0) is true.
		if myHead.Y == boardHeight-1 && isBody(myHead.X, 0, mybody) {
			possibleMoves["up"] = false
		}
		// If our head is at y = 0 then check if isBody(x, boardHeight-1) is true.
		if myHead.Y == 0 && isBody(myHead.X, boardHeight-1, mybody) {
			possibleMoves["down"] = false
		}
	}

	// TODO: Look ahead further than one cell to help with avoiding snakes.
	// Don't collide with others snakes.
	// If another snake is immediately to the right or left of your head, you can't move right or left.
	// Check to see if other snake is to the right of head.
	if isSnake(myHead.X+1, myHead.Y, state.Board.Snakes) {
		possibleMoves["right"] = false
	}
	// Check to see if other snake is to the left of head.
	if isSnake(myHead.X-1, myHead.Y, state.Board.Snakes) {
		possibleMoves["left"] = false
	}
	// If another snake is immediately above or below your head, you can't move up or down.
	// Check to see if other snake is above head.
	if isSnake(myHead.X, myHead.Y+1, state.Board.Snakes) {
		possibleMoves["up"] = false
	}
	// Check to see if other snake is below head.
	if isSnake(myHead.X, myHead.Y-1, state.Board.Snakes) {
		possibleMoves["down"] = false
	}
	// If we are in the wrapped game mode and we are on the edge of the board, we need to avoid wrapping into another snake.
	if gameMode == "wrapped" && onEdge(myHead.X, myHead.Y, state.Board.Width, state.Board.Height) {
		// Print that we are in this section of the code for debugging purposes.
		log.Println("We are in wrapped game mode and don't want to wrap into another snake")
		// If our head is at x = boardWidth - 1 then check if isSnake(0, y) is true.
		if myHead.X == boardWidth-1 && isSnake(0, myHead.Y, state.Board.Snakes) {
			possibleMoves["right"] = false
		}
		// If our head is at x = 0 then check if isSnake(boardWidth-1, y) is true.
		if myHead.X == 0 && isSnake(boardWidth-1, myHead.Y, state.Board.Snakes) {
			possibleMoves["left"] = false
		}
		// If our head is at y = boardHeight - 1 then check if isSnake(x, 0) is true.
		if myHead.Y == boardHeight-1 && isSnake(myHead.X, 0, state.Board.Snakes) {
			possibleMoves["up"] = false
		}
		// If our head is at y = 0 then check if isSnake(x, boardHeight-1) is true.
		if myHead.Y == 0 && isSnake(myHead.X, boardHeight-1, state.Board.Snakes) {
			possibleMoves["down"] = false
		}
	}


	// Avoid Hazards if possible. If the only move is into a hazard, then we will take it.
	// Hazards are represented as a list of coordinates.
	hazards := state.Board.Hazards
	// Only run this code if we are in royale game mode.
	if gameMode == "royale" {
		// Check if neighboring cells are in the list of hazards.
		// If so, you can't move in that direction.
		// Check to see if a hazard is to the left of our head.
		if myHead.X-1 >= 0 {
			if isHazard(myHead.X-1, myHead.Y, hazards) {
				log.Printf("Going to hit a hazard to the left")
				// If two or more safe moves are available, then set left to false.
				if len(safeMoves(possibleMoves)) > 1 {	
					possibleMoves["left"] = false
				}
			}
		}
		// Check to see if hazard is to the right of our head.
		if myHead.X+1 <= state.Board.Width-1 {
			if isHazard(myHead.X+1, myHead.Y, hazards) {
				log.Printf("Going to hit a hazard to the right")
				// If two or more safe moves are available, then set right to false.
				if len(safeMoves(possibleMoves)) > 1 {
					possibleMoves["right"] = false
				}
			}
		}
		// Check to see if hazard is to below our head.
		if myHead.Y-1 >= 0 {
			if isHazard(myHead.X, myHead.Y-1, hazards) {
				log.Printf("Going to hit a hazard below")
				// If two or more safe moves are available, then set down to false.
				if len(safeMoves(possibleMoves)) > 1 {
					possibleMoves["down"] = false
				}
			}
		}
		// Check to see if hazard is above our head.
		if myHead.Y+1 < state.Board.Height-1 {
			if isHazard(myHead.X, myHead.Y+1, hazards) {
				log.Printf("Going to hit a hazard above")
				// If two or more safe moves are available, then set up to false.
				if len(safeMoves(possibleMoves)) > 1 {
					possibleMoves["up"] = false
				}
			}
		}
	}

	// Find food if we are low on health
	// If we are low on health, we want to find food.
	// log the sorted food list
	sortedFood := sortFood(state)
	// If we are low on health, we want to find food.
	if state.Board.Snakes[0].Health < 40 {
	// Go through the sorted food list if we can move there then move there.
		for _, food := range sortedFood {
			if food.X < myHead.X && possibleMoves["left"] {
				log.Printf("Going to eat food to the left")
				// set other moves to false
				possibleMoves["right"] = false
				possibleMoves["up"] = false
				possibleMoves["down"] = false
				break
			} else if food.X > myHead.X && possibleMoves["right"] {
				log.Printf("Going to eat food to the right")
				// set other moves to false
				possibleMoves["left"] = false
				possibleMoves["up"] = false
				possibleMoves["down"] = false
				break
			} else if food.Y < myHead.Y && possibleMoves["down"] {
				log.Printf("Going to eat food below")
				// set other moves to false
				possibleMoves["left"] = false
				possibleMoves["right"] = false
				possibleMoves["up"] = false
				break
			} else if food.Y > myHead.Y && possibleMoves["up"] {
				log.Printf("Going to eat food above")
				// set other moves to false
				possibleMoves["left"] = false
				possibleMoves["right"] = false
				possibleMoves["down"] = false
				break
			}
		}
	}

	// Finally, choose a move from the available safe moves.
	// TODO: Select a move to make based on strategy, rather than random.
	var nextMove string

	if len(safeMoves(possibleMoves)) == 0 {
		nextMove = "down"
		log.Printf("%s MOVE %d: No safe moves detected! Moving %s\n", sanatizeInput(state.Game.ID), isNumber(state.Turn), nextMove)
	} else {
		nextMove = safeMoves(possibleMoves)[rand.Intn(len(safeMoves(possibleMoves)))]
		log.Printf("%s MOVE %d: %s\n", sanatizeInput(state.Game.ID), isNumber(state.Turn), nextMove)
	}
	return BattlesnakeMoveResponse{
		Move: nextMove,
	}
}

File: snakes/python/summer-league-2022/src/logic.py
import random
from typing import List, Dict

from pathfinding.core.diagonal_movement import DiagonalMovement
from pathfinding.core.grid import Grid
from pathfinding.finder.a_star import AStarFinder
from pathfinding.finder.dijkstra import DijkstraFinder

"""
This file can be a nice home for your Battlesnake's logic and helper functions.
We have started this for you, and included some logic to remove your Battlesnake's 'neck'
from the list of possible moves!
"""

def get_info() -> dict:
    """
    This controls your Battlesnake appearance and author permissions.
    For customization options, see https://docs.battlesnake.com/references/personalization
    TIP: If you open your Battlesnake URL in browser you should see this data.
    """
    return {
        "apiversion": "1",
        "author": "es-na-battlesnake",
        "color": "#C5B358",
        "head": "silly",
        "tail": "bolt",
    }

def build_board(board: dict) -> List[List[int]]:
    """
    Create a 2d array the size of the board to represent the board.
    """
    board_map = []
    for row in range(board["height"]):
        board_map.append([])
        for col in range(board["width"]):
            board_map[row].append(2)

    return board_map

def add_snakes_to_board(board: List[List[int]], snakes: List[dict]) -> List[List[int]]:
    """
    board: A 2d array representing the board.
    snakes: A list of dictionaries containing the bodies of the other snakes on the board.
    return: A 2d array representing the board with the other snakes added.
    """
    for snake in snakes:
        for segment in snake["body"]:
            board[segment["y"]][segment["x"]] = 0

    #make our own tail walkable
    board[snakes[0]["body"][0]["y"]][snakes[0]["body"][0]["x"]] = 1
    
    return board

def add_hazards_to_board(board: List[List[int]], hazards: List[dict]) -> List[List[int]]:
    """
    board: A 2d array representing the board.
    hazards: A list of dictionaries containing the bodies of the other snakes on the board.
    return: A 2d array representing the board with the other snakes added.
    """
    for hazard in hazards:
        board[hazard["y"]][hazard["x"]] = -1
    
    return board

def build_grid(board: dict) -> Grid:
    """
    board: A dictionary containing the board information.
    return: A Grid object representing the board.
    """
    grid = Grid(matrix=board)
    return grid

def get_target(data: dict, board: dict) -> dict:
    """
    data: A dictionary containing information about the game.
    return: A dictionary containing the x/y coordinates of the target.
    """
    #print("Head is located at: ", data["you"]["head"]["x"], data["you"]["head"]["y"])
    if data["you"]["head"]["y"] <= 5:
        # print("head is less than 5")

        while True:
            y = random.randint(5, data["board"]["width"]-2)
            x = random.randint(0, data["board"]["height"]-2)

            #print("Target is located at: ", x, y)
            if board[x][y] >= 1:
                break

        return y, x
    else:
        # print("head is greater than 5")

        while True:
            y = random.randint(0, data["board"]["width"]//2)
            x = random.randint(0, data["board"]["height"]-2)

            #print("Target is located at: ", x, y)
            if board[x][y] >= 1:
                break
        
        return y, x

def add_food_to_board(board: List[List[int]], food: List[dict]) -> List[List[int]]:
    """
    board: A 2d array representing the board.
    food: A list of dictionaries containing the bodies of the other snakes on the board.
    return: A 2d array representing the board with the other snakes added.
    """
    for food_item in food:
        board[food_item["y"]][food_item["x"]] = 1
    
    return board

def target_closest_food(data: dict) -> str:
    """
    data: A dictionary containing information about the game.
    return: A string representing the direction to move.
    """
    head = data["you"]["head"]
    food = data["board"]["food"]
    # Find food closest to head
    """
    distance := abs(food.X-state.You.Head.X) + abs(food.Y-state.You.Head.Y)
        // If the distance is less than the closest distance, then set the food to be the closest food.
        if distance < closestDistance || closestDistance == 0 {
            closestDistance = distance
            closestFoodCell = grid.Get(food.X, food.Y)
        }
    """
    for f in food:
        distance = abs(f["x"] - head["x"]) + abs(f["y"] - head["y"])
        closest_distance = 0
        closestFoodCell = 0,0
        if distance < closest_distance or closest_distance == 0:
            closest_distance = distance
            closestFoodCell = f["x"], f["y"]
    
    return closestFoodCell

def get_direction(path: List[GridNode]) -> str:
    """
    path: A list of GridNode objects representing the path to the target.
    return: A string representing the direction to move.
    """

    if len(path) < 2:
        # return random move
        print("Random move")
        return random.choice(["up", "down", "left", "right"])

    current = path[0]
    next_move = path[1]

    if next_move.x == current.x:
        if next_move.y > current.y:
            return "up"
        else:
            return "down"
    elif next_move.y == current.y:
        if next_move.x > current.x:
            return "right"
        else:
            return "left"
    else:
        print("error: no move available")
        return random.choice(["up", "down", "left", "right"])

def choose_move(data: dict) -> str:
    """
    data: Dictionary of all Game Board data as received from the Battlesnake Engine.
    For a full example of 'data', see https://docs.battlesnake.com/references/api/sample-move-request
    return: A String, the single move to make. One of "up", "down", "left" or "right".
    Use the information in 'data' to decide your next move. The 'data' variable can be interacted
    with as a Python Dictionary, and contains all of the information about the Battlesnake board
    for each move of the game.
    """

    snakes = data["board"]["snakes"]
    hazards = data["board"]["hazards"]
    food = data["board"]["food"]
    # Print the move
    print("Move: ", data["turn"])
    board = build_board(data["board"])
    # Add snakes to board
    board = add_snakes_to_board(board, snakes)
    # Add hazards to board
    board = add_hazards_to_board(board, hazards)
    # Add food to board
    board = add_food_to_board(board, data["board"]["food"])
    grid = build_grid(board)

    # if our health is low then target food. Otherwise random target
    if data["you"]["health"] < 80:
        target = target_closest_food(data)   
    else:
        target = get_target(data, board)
    # check if target is walkable. If not get new target

   
    #Set starting point to your head
    start = grid.node(data["you"]["head"]["x"], data["you"]["head"]["y"])
    end = grid.node(target[0], target[1])
    #print("Got a target" , target)
    finder = DijkstraFinder(diagonal_movement=DiagonalMovement.never)
    path, runs = finder.find_path(start, end, grid)

    #print('operations:', runs, 'path length:', len(path))
    #print(grid.grid_str(path=path, start=start, end=end))
    #print(path)

    return get_direction(path)

'''
The functions below this point are only used by the tests. 
They can be removed and tests can be cleaned up.
'''

def _avoid_my_neck(my_body: dict, possible_moves: List[str]) -> List[str]:
    """
    my_body: List of dictionaries of x/y coordinates for every segment of a Battlesnake.
            e.g. [{"x": 0, "y": 0}, {"x": 1, "y": 0}, {"x": 2, "y": 0}]
    possible_moves: List of strings. Moves to pick from.
            e.g. ["up", "down", "left", "right"]
    return: The list of remaining possible_moves, with the 'neck' direction removed
    """

    my_head = my_body[0]  # The first body coordinate is always the head
    my_neck = my_body[1]  # The segment of body right after the head is the 'neck'


    if my_neck["x"] < my_head["x"]:  # my neck is left of my head
        possible_moves.remove("left")
    elif my_neck["x"] > my_head["x"]:  # my neck is right of my head
        possible_moves.remove("right")
    elif my_neck["y"] < my_head["y"]:  # my neck is below my head
        possible_moves.remove("down")
    elif my_neck["y"] > my_head["y"]:  # my neck is above my head
        possible_moves.remove("up")

    return possible_moves

# function to prevent the snake from colliding with itself.
def _avoid_my_body(my_body: dict, possible_moves: List[str]) -> List[str]:

    my_head = my_body[0]

    for segment in my_body:
        if segment["x"] == my_head["x"] and segment["y"] == my_head["y"]:
            continue
        if segment["x"] < my_head["x"]:
            if "left" in possible_moves:
                possible_moves.remove("left")
        elif segment["x"] > my_head["x"]:
            if "right" in possible_moves:
                possible_moves.remove("right")
        elif segment["y"] < my_head["y"]:
            if "down" in possible_moves:
                possible_moves.remove("down")
        elif segment["y"] > my_head["y"]:
            if "up" in possible_moves:
                possible_moves.remove("up")

    return possible_moves

# function to prevent the snake from colliding with walls.
def _avoid_walls(my_body: dict, possible_moves: List[str], board: dict) -> List[str]:

    board_height = board["height"]
    board_width =  board["width"]

    my_head = my_body[0]

    if my_head["x"] + 1 > board_width:
        possible_moves.remove("right")
    if my_head["x"] - 1 < 0:
        possible_moves.remove("left")
    if my_head["y"] + 1 > board_height:
        possible_moves.remove("up")
    if my_head["y"] - 1 < 0:
        possible_moves.remove("down")

    return possible_moves

def _avoid_snake(my_body: dict, other_body: dict, possible_moves: List[str]) -> List[str]:

    my_head = my_body[0]

    # check if head is in the other snake's body. If so, remove the move that would cause collision.
    to_right = {'x': my_head["x"] + 1, 'y': my_head["y"]}
    to_left = {'x': my_head["x"] - 1, 'y': my_head["y"]}
    to_up = {'x': my_head["x"], 'y': my_head["y"] + 1}
    to_down = {'x': my_head["x"], 'y': my_head["y"] - 1}

    for body in other_body:
        if to_right in body:
           possible_moves.remove("right")
        if to_left in body:
            possible_moves.remove("left")
        if to_up in body:
            possible_moves.remove("up")
        if to_down in body:
            possible_moves.remove("down")

    return possible_moves

def _get_other_snakes(data: dict) -> List[dict]:

    """
    data: A dictionary containing information about the game.
    return: A list of dictionaries containing the bodies of the other snakes on the board.
    """
    other_snakes = []
    for snake in data["board"]["snakes"]:
        if snake["id"] != data["you"]["id"]: 
            other_snakes.append([snake])
    return other_snakes
